### PR TITLE
[release/8.0] Infer type mapping suitable for concatenating two results (#32510)

### LINKED ISSUE #32325: LINQ "Contains" generates SQL with wrong char length
## File a bug
My program worked correctly with EF7, but when I upgraded to EF8,I encountered this problem.  When I use Contains(x => x.Property1 + ";" + x.Property2) in a LINQ query, the generated SQL uses a variable length limited to the length of only Property1, so the Contains() always returns zero results.

This also happens with varchar columns, not just char columns.

### Include your code
Full demo attached.
snippet from my attached demo:
```
List<string> queryData = ["AAA;BBBBB", "AAA;CCCCC"];
List<Person> results = dbContext.Persons.Where(x => queryData.Contains(x.ThreeCharacterProperty + ";" + x.FiveCharacterProperty)).ToList();
```
Generated SQL:
```
exec sp_executesql N'SELECT [p].[Id], [p].[FiveCharacterProperty], [p].[ThreeCharacterProperty]
FROM [Persons] AS [p]
WHERE [p].[ThreeCharacterProperty] + '';'' + [p].[FiveCharacterProperty] IN (
    SELECT [q].[value]
    FROM OPENJSON(@__queryData_0) WITH ([value] char(3) ''$'') AS [q]
)',N'@__queryData_0 nvarchar(4000)',@__queryData_0=N'["AAA;BBBBB","AAA;CCCCC"]'

// the problem can be fixed by changing char(3) to char(9) because 9 = 3 + 1 + 5
```
[BugRepro1.zip](https://github.com/dotnet/efcore/files/13383023/BugRepro1.zip)

### Include provider and version information
EF Core version: 8.0.0
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework:.NET 8.0
Operating system: Windows 10 Pro 22H2
IDE: Visual Studio 2022 17.8.0

### PR DESCRIPTION
**Pull request by ajcvickers**
Fixes #32325
Port of #32510

### Description

When using OPENJSON for the new translation of Contains, if the items in the list are concatenated from items that have type facets, then the resultant type mapping might not be big enough to contain the entire value. This then may return incorrect results.

### Customer impact

Queries which previously worked may now contain incorrect results.

### How found

Customers reported on 8.0

### Regression

Yes

### Testing

Added

### Risk

Low and quirked.

### CODE REVIEW

File: src/EFCore.Relational/Query/SqlExpressionFactory.cs [hunk 1]
```diff
@@ -202,6 +205,53 @@ private SqlExpression ApplyTypeMappingOnLike(LikeExpression likeExpression)
             }
 
             case ExpressionType.Add:
+                inferredTypeMapping = typeMapping;
+
+                if (inferredTypeMapping is null && !UseOldBehavior32325)
```
Comment (roji) [inline src/EFCore.Relational/Query/SqlExpressionFactory.cs:210]: Doesn't this mean that if typeMapping is null, we never call ExpressionExtensions.InferTypeMapping as we currently do?

Comment (ajcvickers) [reply src/EFCore.Relational/Query/SqlExpressionFactory.cs]: Fixed.

**Comment by AndriySvyryd** [comment]
Note that the branch is currently closed.

**Review by roji — APPROVED** [review]

**Comment by roji** [comment]
@indrajitjadeja the fix in #32510 only corrected the length on the type mapping; this is a general type mapping inference fix that isn't SQL Server-specific (even if the original bug manifested specifically in SQL Server). I'm still not clear on how it exactly it affects the Oracle SQL translation - can you please open a new issue with a clear minimal, runnable repro and the SQL generated in 8.0.1 and 8.0.2?

**Pull request merged**